### PR TITLE
backupccl: allow user to compress parquet file from EXPORT PARQUET

### PR DESCRIPTION
**Pull request by msbutler**
Previously, EXPORT PARQUET compressed all parquet files with snappy. Now EXPORT
PARQUET defaults to exporting an uncompressed file and provides the
'compression' option which can be either 'gzip' or 'snappy'.

To implement this cleanly, the ParquetWriterProcessor now uses the
CSVWriterSpec, deprecating the ParquetWriterSpec.

Informs #67710

Release note (sql change): EXPORT PARQUET has a new 'compression' option whose
value can be 'gzip' or 'snappy'. An example query:

`EXPORT INTO PARQUET 'nodelocal://0/compress_snappy' WITH compression = snappy
   FROM SELECT * FROM foo `

By default, the parquet file will be uncompressed. With compression, the file
name will be '[filename].parquet.gz' or '[filename].parquet.snappy'.

### CODE REVIEW

File: pkg/sql/execinfrapb/processors_bulk_io.proto [hunk 1]
```diff
@@ -255,22 +255,33 @@ message SplitAndScatterSpec {
...
 // the file name, row count and byte size.
 message CSVWriterSpec {
   // destination as a cloud.ExternalStorage URI pointing to an export store
   // location (directory).
   optional string destination = 1 [(gogoproto.nullable) = false];
   optional string name_pattern = 2 [(gogoproto.nullable) = false];
-  optional roachpb.CSVOptions options = 3 [(gogoproto.nullable) = false];
+  //optional FileFormat file_format = 10 [(gogoproto.nullable) = false];
+  optional roachpb.CSVOptions csv_options = 3 [(gogoproto.nullable) = true];
+  optional roachpb.ParquetOptions parquet_options = 8 [(gogoproto.nullable) = true];
```
Comment (msbutler) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:283]: Now that the Parquet Processor and the CSV Processor use the same kind of spec, I'm wondering if there's any value passing a FileFormat Enum to the spec (a few lines above); or should I stick to my current implementation which maintains the following in the spec: CSVOptions will be not nil IFF the spec is getting passed the CSV Exporter. This same invariant applies to ParquetOptions.

File: pkg/sql/execinfrapb/processors_bulk_io.proto [hunk 2]
```diff
@@ -282,11 +293,15 @@ message CSVWriterSpec {
...
+
+  // col_names specifies the logical column names for the exported parquet file.
+  repeated string col_names = 9 ;
 }
 
 // ParquetWriterSpec is the specification for a processor that consumes rows and
 // writes them to Parquet files at uri. It outputs a row per file written with
 // the file name, row count and byte size.
+// DEPRECATE
 message ParquetWriterSpec {
```
Comment (msbutler) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:305]: Is it safe to deprecate this proto message?

File: pkg/sql/execinfrapb/processors_bulk_io.proto (hunk 1 above)
Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:283]: Should I worry about modifying CSVOptions, a current field in this proto message. Do I need to bump to the processor version in version.go?

File: pkg/sql/execinfrapb/processors_bulk_io.proto [hunk 3]
```diff
@@ -255,22 +255,33 @@ message SplitAndScatterSpec {
...
+// supported for the Exporter spec.
+/*enum FileFormat {
+  Csv = 0;
+  Parquet = 1;
+}*/
+
+// ExporterSpec is the specification for a processor that consumes rows and
+// writes them to Parquet or CSV files at uri. It outputs a row per file written with
 // the file name, row count and byte size.
 message CSVWriterSpec {
```
Comment (msbutler) [inline pkg/sql/execinfrapb/processors_bulk_io.proto:276]: Ok to modify the name of this message to `ExportSpec`?

File: pkg/sql/execinfrapb/processors_bulk_io.proto (hunk 2 above)
Comment (stevendanna) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:305]: I believe we have never shipped a version with this proto message defined so we should be safe to remove it completely.

File: pkg/sql/execinfrapb/processors_bulk_io.proto (hunk 3 above)
Comment (stevendanna) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:276]: I believe that at the gRPC and protofbuf level's renaming a message or field is fine (provided we don't make use of it in protobuf.Any) as long as the type remains the same (or is only different in some compatible way)  The main concern is whether other parts of the system depend on the name. For instance, a rename could break an existing query that makes use of `pb_to_json`.  I've started a discussion in slack about this.

File: pkg/sql/execinfrapb/processors_bulk_io.proto (hunk 1 above)
Comment (stevendanna) [reply pkg/sql/execinfrapb/processors_bulk_io.proto:283]: > I'm wondering if there's any value passing a FileFormat Enum to the spec (a few lines above); or should I stick to my current implementation which maintains the following in the spec: CSVOptions will be not nil IFF the spec is getting passed the CSV Exporter. This same invariant applies to ParquetOptions.

The current implementation seems OK to me. There is some prior art in the existing protobufs that make use of either protobuf's `oneof` or gogoproto.onlyone that might be good to take a peak at if you are looking for alternatives -- although I don't know our typical style here.

Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto]: discovered roachpb.IO_Formats. Just pushed a significant refactor. Conversation no longer applies.

File: pkg/sql/execinfrapb/processors_bulk_io.proto (hunk 3 above)
Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto]: Done

File: pkg/sql/execinfrapb/processors_bulk_io.proto (hunk 2 above)
Comment (msbutler) [reply pkg/sql/execinfrapb/processors_bulk_io.proto]: Done

**Comment by msbutler** [comment]
added @yuzefovich as a reviewer to make sure that the dist sql version doesn't need to be bumped given my changes to the processor spec.

File: pkg/sql/execinfrapb/processors.proto [hunk 4]
```diff
@@ -101,7 +101,7 @@ message ProcessorCoreUnion {
   optional BackfillerSpec backfiller = 11;
   optional ReadImportDataSpec readImport = 13;
   reserved 14;
-  optional CSVWriterSpec CSVWriter = 20;
+  optional ExportSpec exporter = 20;
```
Comment (msbutler) [inline pkg/sql/execinfrapb/processors.proto:104]: @yuzefovich perhaps its just safer to push ExportSpec to a new processor number?

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 6 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74661)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @msbutler, and @stevendanna)

___
*[pkg/sql/execinfrapb/processors.proto, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74661#-Mtyc8mmcoc-ULcT57-N-r2-104:-MuDxhcs3O5qwyjumV5q:b-t8emz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbdc19252df4de0aec8d187e1951a7d1a5bd7d39/pkg/sql/execinfrapb/processors.proto#L104)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

@yuzefovich perhaps its just safer to push ExportSpec to a new processor number and retire processor spec 20?
</blockquote></details>

I think it is safe to keep using the same number and simply rename the protobuf message (we do need to add a couple of `reserved` things I mention later), but it might be cleaner to just get a new number for `ExportSpec` and retire the number 20. We've already dumped the DistSQL version multiple times throughout this development cycle, so trying not to do it for `ExportSpec` won't give us any benefit.

In short, it's really up to you; personally, I'd probably bump the DistSQL version.
___
*[pkg/sql/execinfrapb/processors_bulk_io.proto, line 273 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74661#-MuDxLEjBnlzsxIp1GMm:-MuDxLEjBnlzsxIp1GMn:bt8rv2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1d95fe05f27dfcea28b1cb4b54e106dd5126078/pkg/sql/execinfrapb/processors_bulk_io.proto#L273)):*
> ```protobuf
>   optional string destination = 1 [(gogoproto.nullable) = false];
>   optional string name_pattern = 2 [(gogoproto.nullable) = false];
>   optional roachpb.CSVOptions options = 3 [(gogoproto.nullable) = false];
> ```

If we're keeping the same number for `ExportSpec` as we had for `CSVWriterSpec`, then we need to mark `3` as `reserved` here. Ditto for `5`.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74661)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @stevendanna, and @yuzefovich)

___
*[pkg/sql/execinfrapb/processors.proto, line 104 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74661#-Mtyc8mmcoc-ULcT57-N-r2-104:-MuRsxZu9Mfi2x4cKEuf:b-7rp00q) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbdc19252df4de0aec8d187e1951a7d1a5bd7d39/pkg/sql/execinfrapb/processors.proto#L104)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think it is safe to keep using the same number and simply rename the protobuf message (we do need to add a couple of `reserved` things I mention later), but it might be cleaner to just get a new number for `ExportSpec` and retire the number 20. We've already dumped the DistSQL version multiple times throughout this development cycle, so trying not to do it for `ExportSpec` won't give us any benefit.

In short, it's really up to you; personally, I'd probably bump the DistSQL version.
</blockquote></details>

decided to bump. done
___
*[pkg/sql/execinfrapb/processors_bulk_io.proto, line 273 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74661#-MuDxLEjBnlzsxIp1GMm:-MuRsv8I0keYNgvLP-TX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1d95fe05f27dfcea28b1cb4b54e106dd5126078/pkg/sql/execinfrapb/processors_bulk_io.proto#L273)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

If we're keeping the same number for `ExportSpec` as we had for `CSVWriterSpec`, then we need to mark `3` as `reserved` here. Ditto for `5`.
</blockquote></details>

Done.

**Comment by msbutler** [comment]
Hmmm. I'm not sure why `TestCSVImportCanBeResumed` is failing under `testrace` (locally and in CI). [CI claims ](https://teamcity.cockroachdb.com/viewLog.html?buildId=4179807&tab=buildResultsDiv&buildTypeId=Cockroach_UnitTests&logTab=)the test didn't fail on my previous commit. But now, when I check out the previous [commit](https://github.com/cockroachdb/cockroach/commit/dbdc19252df4de0aec8d187e1951a7d1a5bd7d39) or even on master, and rerun the test locally, it fails consistently with the following stacktrace. 
```
e5
github.com/cockroachdb/cockroach/pkg/util/log.init()
        /Users/michaelbutler/go/src/github.com/cockroachdb/cockroach/pkg/util/log/file_names.go:71 +0x1ca fp=0xc00035f968 sp=0xc00035f6f8 pc=0x4d74bea
runtime.doInit(0x106bf440)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6498 +0x123 fp=0xc00035faa0 sp=0xc00035f968 pc=0x404af03
runtime.doInit(0x106b49a0)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6475 +0x71 fp=0xc00035fbd8 sp=0xc00035faa0 pc=0x404ae51
runtime.doInit(0x106ad9a0)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6475 +0x71 fp=0xc00035fd10 sp=0xc00035fbd8 pc=0x404ae51
runtime.doInit(0x106dbba0)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6475 +0x71 fp=0xc00035fe48 sp=0xc00035fd10 pc=0x404ae51
runtime.doInit(0x10695880)
        /usr/local/opt/go/libexec/src/runtime/proc.go:6475 +0x71 fp=0xc00035ff80 sp=0xc00035fe48 pc=0x404ae51
runtime.main()
        /usr/local/opt/go/libexec/src/runtime/proc.go:238 +0x1e6 fp=0xc00035ffe0 sp=0xc00035ff80 pc=0x403d826
runtime.goexit()
        /usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc00035ffe8 sp=0xc00035ffe0 pc=0x4070e01

rax    0x0
rbx    0x1a3bb600
rcx    0x7ff7bfefed68
rdx    0x0
rdi    0x103
rsi    0x6
rbp    0x7ff7bfefed90
rsp    0x7ff7bfefed68
r8     0x0
r9     0x0
r10    0x0
r11    0x246
r12    0x103
r13    0x50
r14    0x6
r15    0x16
rip    0x7ff800a87112
rflags 0x246
cs     0x7
fs     0x0
gs     0x0
FAIL    github.com/cockroachdb/cockroach/pkg/ccl/importccl      2.587s
FAIL
make: *** [Makefile:1081: testrace] Error 1
```

**Comment by stevendanna** [comment]
@msbutler Interest, your stacktrace there looks more like the problems I've seen people hit after a macOS upgrade when they have a lot of cached go artifacts built on a previous release.  I haven't dug into that error, but have resolved it in the past by making sure I had an up-to-date Go version, doing a make clean, and rebuilding things from the ground up (which is a bummer and not very principled).

Poking at the logs in CI, I see a lot of:

```
E220128 15:40:36.474308 434533 kv/kvserver/queue.go:1095 ⋮ [n1,mvccGC,s1,r2/1:‹/System/NodeLiveness{-Max}›] 111  cannot ‹read› undeclared span ‹/{Min-System/NodeLiveness/1}›
E220128 15:40:36.474308 434533 kv/kvserver/queue.go:1095 ⋮ [n1,mvccGC,s1,r2/1:‹/System/NodeLiveness{-Max}›] 111 +declared:
E220128 15:40:36.474308 434533 kv/kvserver/queue.go:1095 ⋮ [n1,mvccGC,s1,r2/1:‹/System/NodeLiveness{-Max}›] 111 +‹read local: /Local/Range/System/NodeLiveness/RangeDescriptor at 0,0›
E220128 15:40:36.474308 434533 kv/kvserver/queue.go:1095 ⋮ [n1,mvccGC,s1,r2/1:‹/System/NodeLiveness{-Max}›] 111 +‹read local: /Local/Lock/Intent/Local/Range/System/NodeLiveness/RangeDescriptor at 0,0›
E220128 15:40:36.474308 434533 kv/kvserver/queue.go:1095 ⋮ [n1,mvccGC,s1,r2/1:‹/System/NodeLiveness{-Max}›] 111 +‹write global: /System/NodeLiveness/1 at 1643384436.406046280,0›
E220128 15:40:36.474308 434533 kv/kvserver/queue.go:1095 ⋮ [n1,mvccGC,s1,r2/1:‹/System/NodeLiveness{-Max}›] 111 +‹write local: /Local/Lock/Intent/System/NodeLiveness/1 at 0,0›
E220128 15:40:36.474308 434533 kv/kvserver/queue.go:1095 ⋮ [n1,mvccGC,s1,r2/1:‹/System/NodeLiveness{-Max}›] 111 +stack:
```

**Comment by msbutler** [comment]
@stevendanna github CI finally passed! I think you were right about updating go and cleaning the old binaries after the OS update. I guess some hard to track build artifacts from my local failed testrace trickled into CI some how? Please review away! Once this merges, I'll open the last big export Parquet PR.

File: pkg/sql/execinfra/version.go [hunk 5]
```diff
@@ -39,18 +39,21 @@ import "github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 //
 // ATTENTION: When updating these fields, add a brief description of what
 // changed to the version history below.
-const Version execinfrapb.DistSQLVersion = 59
+const Version execinfrapb.DistSQLVersion = 60
 
 // MinAcceptedVersion is the oldest version that the server is compatible with.
 // A server will not accept flows with older versions.
-const MinAcceptedVersion execinfrapb.DistSQLVersion = 58
+const MinAcceptedVersion execinfrapb.DistSQLVersion = 60
```
Comment (stevendanna) [inline pkg/sql/execinfra/version.go:46]: Just for reference: on release-21.2 the MinAcceptedVersion was `48` so it was already the case that there was going to be no window of compatibility between 21.2 and 22.1 nodes. 
https://github.com/cockroachdb/cockroach/blob/release-21.2/pkg/sql/execinfra/version.go#L40-L46

**Review by stevendanna — APPROVED** [review]
This definitely seems a lot cleaner than to me.  Let's be sure to double check our mixed-version import/export tests during stability just in case.

**Comment by msbutler** [comment]
bors r = stevendanna

**Comment by msbutler** [comment]
bors r=stevendanna

**Comment by msbutler** [comment]
TFTR !
